### PR TITLE
fix: topkgating major bug

### PR DESCRIPTION
**Pull request by excepshenal**
`topk_masked_gates` was previously being used across the tokens dimension to determine which tokens had the highest importance for each expert. However, it was using logits rather than probabilities to determine this.

This was causing print statements like
```
            if dist.get_rank() == 0:
                print(f"Mask mean: {mask.float().mean()}")
                print(f"Capacity mask mean: {capacity_mask.mean()}")
            mask = torch.logical_and(mask, capacity_mask)
            if dist.get_rank() == 0:
                print(f"Mask (after AND) mean: {mask.float().mean()}")
```

to often yield values like
```
Mask mean: 0.0625
Capacity mask mean: 0.0625
Mask (after AND) mean: 0.005908316932618618
```

and in turn the average number of routed experts per token was as low as `0.001`.

### CODE REVIEW

File: deepspeed/moe/sharded_moe.py [hunk 1]
```diff
@@ -384,14 +384,13 @@ def topkgating(
...
-    top_gate, top_idx = torch.topk(logits, k=k, dim=1)
     # gating decisions
     gates = F.softmax(logits, dim=1)
     num_experts = int(gates.shape[1])
 
-    # get topk mask
-    topk_masked_gates = torch.zeros_like(logits).scatter(1, top_idx, top_gate)
+    # get topk gates — use softmax probs so non-assigned slots (zero) never
+    # outrank legitimately-assigned tokens that happen to have negative logits
+    top_gate, top_idx = torch.topk(gates, k=k, dim=1)
```
[the remark this answers is not in the document]
Comment (delock) [reply deepspeed/moe/sharded_moe.py:392]: @excepshenal should it be better if topk are picked from logits (so there is no concern of underflow) and topk_masked_gates are computed from probabilities?  We can use logits here and still get `(top_gate, top_idx)` tuple as we expected, right?

**Review by delock — CHANGES_REQUESTED** [review]

File: deepspeed/moe/sharded_moe.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (excepshenal) [reply deepspeed/moe/sharded_moe.py:392]: We could, at the cost of another `torch.topk` call, when I don't think there is any benefit because when the gate probability is 0 for an expert, it won't contribute to the final residual anyway, so picking among such experts won't affect the calculation.

For reference, HuggingFace and Megatron also run `topk` on softmax probs not logits.

**Review by delock — APPROVED** [review]

**Comment by delock** [comment]
Hi @excepshenal ，can you fix DCO error by sign-off your commits?  Thanks!
